File: tfcc_runtime/operations/base/at1.cpp
// Copyright 2021 Wechat Group, Tencent
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "at1.h"

#include <algorithm>

#include "tfcc_runtime/framework/graph.h"
#include "tfcc_runtime/proto/operations/base.pb.h"
#include "tfcc_runtime/utils/commutils.h"

namespace tfcc {
namespace runtime {
namespace base {

template <class T>
static const char* _wrapper_at1(const std::vector<T>* a, size_t pos, T* result) noexcept {
  if (pos >= a->size()) {
    return "At1: invalid pos";
  }
  *result = (*a)[pos];
  return nullptr;
}

template <tfcc::runtime::common::DataType DTYPE>
std::string At1<DTYPE>::getOperationTypeUrl() const {
  tfcc::runtime::operations::base::At1 operation;
  return get_protobuf_type_url(operation);
}

template <tfcc::runtime::common::DataType DTYPE>
std::set<unsigned> At1<DTYPE>::getOperationVersions() const {
  std::set<unsigned> versions = {tfcc::runtime::operations::base::At1::VERSION_1};
  return versions;
}

template <tfcc::runtime::common::DataType DTYPE>
std::unique_ptr<OperationResource> At1<DTYPE>::process(
    const tfcc::runtime::model::Node& node, Graph& graph) const {
  if (node.inputs_size() != 1 || node.outputs_size() != 1) {
    return nullptr;
  }
  if (!node.operation().Is<tfcc::runtime::operations::base::At1>()) {
    return nullptr;
  }
  auto& symbolManager = getSymbolManager(graph);
  if (symbolManager.getSymbolInfo(node.outputs(0)).dtype != DTYPE ||
      symbolManager.getSymbolInfo(node.inputs(0)).dtype != DTYPE) {
    return nullptr;
  }

  std::unique_ptr<OperationResource> resource(new OperationResource);
  tfcc::runtime::operations::base::At1 operation;
  node.operation().UnpackTo(&operation);

  auto& jit = getJIT(graph);

  const std::vector<T>* inputSymbol = symbolManager.getVector(node.inputs(0), T());
  T* outputSymbol = symbolManager.getValue(node.outputs(0), T());

  callFunction(jit, _wrapper_at1<T>, inputSymbol, operation.idx(), outputSymbol);

  return resource;
}

std::vector<std::unique_ptr<Operation>> get_at1_operations() {
#define DEFINE_FUNC(dtype) operations.emplace_back(std::unique_ptr<Operation>(new At1<dtype>()));
  std::vector<std::unique_ptr<Operation>> operations;
  TFCC_RUNTIME_FOR_ALL_DATA_TYPE(DEFINE_FUNC);
  return operations;
}

}  // namespace base
}  // namespace runtime
}  // namespace tfcc

File: tfcc_code_generator/ir/node/base/at1.py
# Copyright 2021 Wechat Group, Tencent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from ir.node import Node
import ir.framework


class At1(Node):
    def update_attributes(self, idx):
        self._idx = idx
        assert idx >= 0
        assert isinstance(idx, int)

    def inference(self):
        assert len(self.inputs) == 1
        assert len(self.outputs) == 1
        assert self.inputs[0].is_vector()

        self.outputs[0].stype = ir.framework.SymbolType.VALUE
        self.outputs[0].dtype = self.inputs[0].dtype
        self.outputs[0].shape = [1]

    @property
    def idx(self):
        return self._idx

    def calculatable(self):
        if super().calculatable():
            return True
        if isinstance(self.inputs[0].incomplete_data, list) and isinstance(
            self.inputs[0].incomplete_data[self.idx], (int, float)
        ):
            return True
        return False

    def calculate(self):
        self.outputs[0].origin_stype = self.outputs[0].stype
        self.outputs[0].stype = ir.framework.SymbolType.CONSTANT_VALUE
        if self.inputs[0].is_constant():
            self.outputs[0].data = np.asarray(
                [self.inputs[0].data.tolist()[self.idx]],
                dtype=self.inputs[0].dtype.numpy_dtype,
            )
        elif isinstance(self.inputs[0].incomplete_data[self.idx], (int, float)):
            self.outputs[0].data = np.asarray(
                [self.inputs[0].incomplete_data[self.idx]],
                dtype=self.inputs[0].dtype.numpy_dtype,
            )
        else:
            raise RuntimeError("Unknow error")

    def calculate_incomplete_data(self):
        if isinstance(self.inputs[0].incomplete_data, list):
            self.outputs[0].incomplete_data = [self.inputs[0].incomplete_data[self.idx]]

    @property
    def attributes(self):
        return {"idx": self.idx}
